### PR TITLE
sdr-sat,sdr-ui: satellites scheduler UI + ZIP/elevation lookup (closes #481)

### LINKED ISSUE #481: Satellite pass scheduler UI (new activity tab)
## Scope

User-facing panel for the satellite-pass predictor (#480). Either a new activity-bar tab ("Satellites") or a panel inside an existing activity — TBD during design.

## Responsibilities

- Display ground-station settings (lat/lon, elevation) with edit
- Show next N passes (default 6) with countdown, satellite name, max elevation, direction
- Highlight current pass if one is in progress
- TLE last-updated indicator + manual refresh button
- "Auto-record this pass" toggle (drives the scheduler wiring ticket #TBD)

## Layout

Standard `AdwPreferencesPage` with three groups:

1. **Ground station** — lat / lon / elevation spin rows, with "Use GPS" / "Enter manually" options. First-run defaults to some reasonable central-US value; user adjusts.
2. **TLE status** — last-refreshed timestamp, manual refresh button, per-satellite freshness indicator.
3. **Next passes** — expandable rows, one per upcoming pass. Each row title is the satellite name + countdown; subtitle shows elevation + direction.

## Persistence

Config keys:
- `sat_station_lat_deg`, `sat_station_lon_deg`, `sat_station_alt_m`
- `sat_tle_last_refresh` (RFC3339 timestamp)
- `sat_auto_record_apt` (bool)

## Notes

- Activity-bar integration follows the existing pattern documented in `CLAUDE.md` (add to `LEFT_ACTIVITIES` slice in `activity_bar.rs`)
- Probably want a satellite emoji / icon — 🛰 or `network-wireless-symbolic`
- Live countdown updates via a 1-second glib timer

## Out of scope
- Actually recording passes (see next ticket)
- Live azimuth/elevation radar plot (v2 polish)

Depends on: #480

Part of epic #468.

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

- Adds the **Satellites** left-sidebar activity (Ctrl+7) with a four-group panel: Ground Station / TLE Data / Recording / Upcoming Passes. 1 Hz countdown ticker keeps each pass row's title in sync ("NOAA 19 — in 12 min" → "starting now" → "in progress (3 min in)").
- **Refactors `sdr-sat::TleCache` to per-NORAD fetches.** Celestrak deprecated the per-group `.txt` URLs (`noaa.txt` is now 404; `noaa` group slug is gone from the GP API; NOAA 15/18/19 aren't in any current group). New endpoint is `gp.php?CATNR={id}&FORMAT=tle`. Cache files become `~/.cache/sdr-rs/tle/{norad_id}.tle`. Validation gate, atomic-rename writes, fetcher injection for tests, and stale-cache fallback all carry over.
- **Two small blocking-HTTP helpers in `sdr-sat`:** `lookup_us_zip` (Zippopotam.us → lat/lon) and `lookup_elevation_m` (Open-Elevation → altitude). The panel chains them on a single worker hop so a US ZIP fills in lat/lon/alt in one click.

## Why ZIP instead of OS location services

GeoClue2 needs the app to be in `/etc/geoclue/geoclue.conf`'s allowlist (typically only flatpak/snap apps) — would fail silently for cargo-installed users. ZIP lookup gives ~1 km accuracy (vs IP geolocation's ~50 km), works without platform-specific permission UIs, and ports cleanly to the planned mac/Windows front-ends. US-only by design for V1; international users enter coords by hand.

## Why per-NORAD instead of group files

Per-NORAD dodges the group-churn problem entirely and lets `KNOWN_SATELLITES` grow without anyone having to figure out which Celestrak group each satellite lives in.

## Test plan

- [x] `cargo build --workspace --features whisper-cpu` clean
- [x] `cargo test --workspace --features whisper-cpu` — 59 sdr-sat tests pass (incl. 14 new postal/elevation tests + new URL-shape pin)
- [x] `cargo clippy --workspace --features whisper-cpu --all-targets -- -D warnings` clean
- [x] `cargo fmt --all --check` clean
- [x] `cargo check --workspace --no-default-features --features sherpa-cpu` clean (transcription not touched but per the dual-build rule)
- [x] Manual smoke (live NOAA 19 pass): activity bar entry + Ctrl+7 toggles correctly; ZIP lookup populates lat/lon/alt with real ground elevation; TLE refresh produces 7 `{norad_id}.tle` files in cache dir; upcoming passes recompute on lat/lon edit; 1 Hz countdown updates rows in place.
- [ ] CodeRabbit review

## Related

- Part of epic #468 (NOAA APT)
- Closes #481
- Unblocks #482 — the auto-record-on-pass UX, which now has a clearer scope including the decoder-wiring precursor (audio → `AptDecoder` → `AptImage` → viewer). Updated #482 body to reflect that.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a Satellites sidebar panel (Ctrl+7) for ground‑station setup and pass scheduling.
  * ZIP-to-location lookup and automatic ground‑altitude lookup for station inputs with status feedback.
  * Per‑satellite TLE caching with manual refresh, “Last refreshed” status, and graceful UI when cache is unavailable.
  * Real‑time pass enumeration and 1 Hz countdown updates, auto‑record APT toggle, and persisted station/TLE settings.

### CODE REVIEW

File: crates/sdr-ui/src/sidebar/satellites_panel.rs [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+        let m = (secs % 3_600) / 60;
+        format!("in {h}h {m:02}m")
+    } else if secs > 60 {
+        format!("in {} min", secs / 60)
+    } else if secs > 0 {
+        "starting now".to_string()
+    } else if to_end.num_seconds() > 0 {
+        format!("in progress ({} min in)", -secs / 60)
+    } else {
+        "ended".to_string()
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/satellites_panel.rs]: Fixed in 851ed05.

* Hoisted `SECS_PER_MINUTE` / `SECS_PER_HOUR` as named constants per CLAUDE.md "name your magic numbers".
* Switched both threshold checks to `>=`: `secs >= SECS_PER_HOUR` (so 3600s reads "in 1h 00m" not "in 60 min") and `secs >= SECS_PER_MINUTE` (so 60s reads "in 1 min" not "starting now").
* Clamped the in-progress min count with `.max(1)` so the first 60 s of an active pass reads "in progress (1 min in)" rather than "0 min in".

Three new unit tests pin all three boundary cases:

* `format_pass_title_at_exact_one_hour_uses_h_m_format`
* `format_pass_title_at_exact_one_minute_says_one_min`
* `format_pass_title_clamps_in_progress_min_to_at_least_one`

File: crates/sdr-ui/src/window.rs [hunk 2]
```diff
@@ -8293,6 +8295,386 @@ fn connect_scanner_panel(
...
+        for pass in passes {
+            let row = adw::ActionRow::builder()
+                .title(format_pass_title(&pass, now))
+                .subtitle(format_pass_subtitle(&pass))
+                .build();
+            panel_recompute.passes_group.add(&row);
+            new_rows.push(DisplayedPass { row, pass });
+        }
+        *displayed_recompute.borrow_mut() = new_rows;
+    });
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 851ed05. You're right — the cycle was real and the timer's `WeakRef::upgrade` exit gate would never fire because the panel widgets couldn't drop.

Refactor:

* Added `SatellitesPanelWeak` (parallel struct of `glib::WeakRef<T>` fields, one per widget) plus `SatellitesPanel::downgrade()` and `SatellitesPanelWeak::upgrade() -> Option<SatellitesPanel>` that does an atomic upgrade — returns `None` the moment any single field's underlying GObject has been dropped, so each closure can short-circuit cleanly with one `let-else`.
* Refactored every closure in `connect_satellites_panel` to capture `panel_weak.clone()` (a struct of cheap weak-ref clones) and upgrade-or-return at the top instead of cloning the strong `SatellitesPanel`. Touched: the `recompute` `Rc<dyn Fn()>`, the refresh-button click handler (and its inner `glib::spawn_future_local` continuation), the ZIP-row apply handler (and its inner continuation), and the 1 Hz countdown timer.
* The 1 Hz timer's exit gate now reads `panel_weak_tick.upgrade().is_none() → ControlFlow::Break`, so it actually fires when the panel drops — which it can do now, with the strong-ref cycle gone.

Followed the same `*Weak` struct-pair pattern referenced from PRs #335 / #375 (per project history on similar widget-handler wiring).

File: crates/sdr-ui/src/window.rs [hunk 3]
```diff
@@ -8293,6 +8295,386 @@ fn connect_scanner_panel(
...
+    let cache = match TleCache::new() {
+        Ok(c) => std::sync::Arc::new(c),
+        Err(e) => {
+            tracing::warn!("Satellites panel: TLE cache unavailable — {e}");
+            panel.refresh_button.set_sensitive(false);
+            panel
+                .last_refresh_row
+                .set_subtitle("Cache directory unavailable");
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 851ed05.

Changed `cache` from `Arc<TleCache>` to `Option<Arc<TleCache>>`. On `TleCache::new()` failure we now log + disable the refresh button + set the "Cache directory unavailable" subtitle, but no longer return out of the function. Wiring that depends on the cache (refresh-button click handler, the real `recompute` body, the 1 Hz countdown timer) is gated on `cache.is_some()`; everything else stays wired:

* lat/lon/alt `value-notify` handlers — still persist to config.
* Auto-record `active-notify` handler — still persists to config (so #482's read picks it up regardless of cache state).
* ZIP-code apply handler — still runs the Zippopotam.us + Open-Elevation lookups; populates lat/lon/alt as before.

`recompute` is built as a no-op `Rc::new(|| {})` when the cache is `None`, so the lat/lon/alt notify handlers can call it unconditionally — keeps the call sites branch-free.

File: crates/sdr-ui/src/sidebar/satellites_panel.rs [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+            Err(TleCacheError::NotFound { .. }) => {
+                // Common case before the first refresh — quiet log.
+                tracing::debug!(
+                    "no cached TLE for {} (NORAD {}); refresh to enable",
+                    known.name,
+                    known.norad_id,
+                );
+            }
+            Err(e) => log_satellite_skip(known.name, &e.to_string()),
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/satellites_panel.rs:594]: Fixed in 25d0703.

Added `TleCache::cached_tle_for(norad_id)` — a strict read-only lookup that never touches the network regardless of cache freshness. Switched `enumerate_upcoming_passes` to use it instead of `tle_for`, so a lat/lon/alt edit on the GTK main thread can no longer block on a synchronous HTTP fetch.

Three new hermetic tests pin the contract:
* `cached_tle_for_does_not_call_network_on_miss` — `NotFound` when no cache file is present, with no fetcher invocation.
* `cached_tle_for_returns_pair_when_file_is_present` — happy path.
* `cached_tle_for_never_invokes_the_fetcher_even_when_stale` — the load-bearing one; counts fetcher calls via an `AtomicU32` to prove the cache-only contract holds even when the cache is past its freshness window.
* `cached_tle_for_returns_not_found_for_corrupted_body` — validation gate applies on the read path too (so a stray HTML response in the cache surfaces as `NotFound`, not garbage).

File: crates/sdr-ui/src/window.rs [hunk 5]
```diff
@@ -8293,6 +8295,424 @@ fn connect_scanner_panel(
...
+                panel.refresh_spinner.set_visible(false);
+                panel.refresh_button.set_sensitive(true);
+
+                match result {
+                    Ok(Ok(())) => {
+                        let now = chrono::Utc::now();
+                        save_tle_last_refresh(&config_done, now);
+                        panel
+                            .last_refresh_row
+                            .set_subtitle(&format_last_refresh(&config_done));
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:8564]: Fixed in 25d0703.

Added `TleCache::force_refresh(norad_id)` — always invokes the fetcher, validates the response, writes to cache, and (critically) does NOT fall back to a stale cache on failure. So the manual Refresh button can only mark "Last refreshed: now" when an actual network round-trip succeeded.

Switched the refresh-button click handler in `connect_satellites_panel` to call `force_refresh` per known-satellite. Two new tests pin the contract:
* `force_refresh_always_calls_fetcher_even_when_cache_is_fresh` — drops a fresh cache file, calls `tle_text` once (no fetcher invocation), then `force_refresh` twice and asserts call count = 2 via an `AtomicU32`.
* `force_refresh_does_not_fall_back_to_stale_cache_on_failure` — drops a valid stale cache, configures a canned-fail fetcher, asserts `force_refresh` returns `Err(Fetch)` rather than silently serving stale data.

File: crates/sdr-ui/src/window.rs [hunk 6]
```diff
@@ -8293,6 +8295,424 @@ fn connect_scanner_panel(
...
+                    let loc = sdr_sat::lookup_us_zip(&zip_for_task)?;
+                    let elevation = match sdr_sat::lookup_elevation_m(loc.lat_deg, loc.lon_deg)
+                    {
+                        Ok(m) => Some(m),
+                        Err(e) => {
+                            tracing::warn!(
+                                "elevation lookup at ({lat:.4}, {lon:.4}) failed: {e}",
+                                lat = loc.lat_deg,
+                                lon = loc.lon_deg,
+                            );
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 25d0703.

Stripped user-location data from both warn lines. Now:

* `tracing::warn!("ZIP lookup failed: {e}");` — no ZIP digits.
* `tracing::warn!("elevation lookup failed: {e}");` — no lat/lon coordinates.

The provider error message alone is enough for debugging; the failure is already surfaced inline in the status row, so we're not losing user-facing information by trimming the log.

File: crates/sdr-ui/src/sidebar/satellites_panel.rs [hunk 7]
```diff
@@ -0,0 +1,808 @@
...
+        let cfg = Arc::new(ConfigManager::in_memory(&serde_json::json!({
+            "sat_tle_last_refresh": "2024-06-15T18:30:00Z",
+        })));
+        let formatted = format_last_refresh(&cfg);
+        assert_ne!(formatted, "Never");
+        assert!(
+            !formatted.contains('T'),
+            "RFC3339 'T' separator leaked into user-facing format: {formatted}"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/satellites_panel.rs:703]: Fixed in 25d0703.

Replaced the `!formatted.contains('T')` check (which would false-fail in any timezone whose abbreviation contains a `T` — `UTC`, `EDT`, `PDT`, etc.) with a position-10 separator check:

```rust
assert_eq!(
    formatted.chars().nth(10),
    Some(' '),
    "expected a space at position 10 (date/time separator); got: {formatted}",
);
```

Pins the same invariant ("output is the local-formatted form, not raw RFC3339") because RFC3339 has `T` at byte 10 and our `%Y-%m-%d %H:%M %Z` format has a space there. Timezone-independent.

File: crates/sdr-ui/src/window.rs [hunk 8]
```diff
@@ -8293,6 +8295,454 @@ fn connect_scanner_panel(
...
+                    return;
+                }
+                let Some(panel) = panel_weak_zip.upgrade() else {
+                    return;
+                };
+                let zip = entry.text().to_string();
+                if zip.trim().is_empty() {
+                    // Empty entry — nothing to do; treat as a no-op so a
+                    // stray Enter doesn't reset the status row.
+                    return;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 928dcc2.

Trim is now done at the read site:

```rust
let zip = entry.text().trim().to_string();
if zip.is_empty() {
    return;
}
```

So the trimmed value is what flows through the rest of the handler — the `length=N` debug log reports the right length, and `lookup_us_zip` (which still trims internally as a defensive belt-and-braces) gets a pre-cleaned input.

File: crates/sdr-ui/src/window.rs [hunk 9]
```diff
@@ -8293,6 +8295,454 @@ fn connect_scanner_panel(
...
+    // the timer would tick uselessly. Captures the panel weakly
+    // so the source returns `ControlFlow::Break` once any panel
+    // widget has been dropped (otherwise GLib runs it forever,
+    // holding a strong chain into the `displayed` vec and its
+    // widgets).
+    if cache.is_some() {
+        let panel_weak_tick = panel_weak.clone();
+        let displayed_tick = Rc::clone(&displayed);
+        let recompute_tick = Rc::clone(&recompute);
+        let _ = glib::timeout_add_local(Duration::from_secs(1), move || {
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 928dcc2.

Hoisted `SATELLITES_COUNTDOWN_TICK: Duration = Duration::from_secs(1)` near the top of `window.rs` next to `DSP_POLL_INTERVAL_MS`, with a comment explaining why 1 s is the natural cadence (smaller would burn cycles for no visible UI change since the pass-row title renders at 1-min granularity for far passes).

```rust
let _ = glib::timeout_add_local(SATELLITES_COUNTDOWN_TICK, move || { ... });
```

File: crates/sdr-ui/src/window.rs [hunk 10]
```diff
@@ -8293,6 +8302,459 @@ fn connect_scanner_panel(
...
+                            };
+                            panel.zip_status_row.set_title(&status);
+                        }
+                        Ok(Err(e)) => {
+                            // Don't include the ZIP in the log — user
+                            // location data, already surfaced inline in
+                            // the status row. Provider error alone is
+                            // enough.
+                            tracing::warn!("ZIP lookup failed: {e}");
+                            panel.zip_status_row.set_title(&e.to_string());
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:8706]: Fixed in fe72961.

Changed the spawn_blocking return type's elevation slot from `Option<f64>` to `Result<f64, String>`. The worker thread converts `ElevationLookupError` to `String` at the boundary — cheap to send across the thread, decoupled from the error type, and the upstream error message itself contains no user-location data (it's HTTP status / parse error / dataset-coverage text).

On a failed elevation lookup the status row now reads e.g.:

> Resolved: Christiansburg, VA (altitude unchanged: elevation lookup HTTP error: HTTP status: 502 Bad Gateway)

instead of the opaque "(altitude unchanged)". Logs are still scrubbed of lat/lon — only the provider error string ends up there. Matches the existing precedent for ZIP failures (which already surface the provider error inline).
